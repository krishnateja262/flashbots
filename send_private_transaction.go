package flashbots

import (
	"encoding/json"
	"math/big"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/common/hexutil"
	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/rpc"
	"github.com/lmittmann/w3/core"
)

type SendPrivateTransactionRequest struct {
	Transaction    *types.Transaction // Signed transaction to send.
	RawTransaction []byte             // Raw signed transaction to send.
	MaxBlockNumber *big.Int           // Max block number for which the tx should be included (Optional).
	Fast           bool               // Enable fast mode (Optional). See https://docs.flashbots.net/flashbots-protect/rpc/fast-mode
}

type sendPrivateTransactionRequest struct {
	RawTransaction hexutil.Bytes `json:"tx"`
	MaxBlockNumber *hexutil.Big  `json:"maxBlockNumber"`
	Preferences    struct {
		Fast bool `json:"fast"`
	} `json:"preferences"`
}

// MarshalJSON implements the json.Marshaler interface.
func (c SendPrivateTransactionRequest) MarshalJSON() ([]byte, error) {
	var enc sendPrivateTransactionRequest

	if c.Transaction != nil {
		rawTx, err := c.Transaction.MarshalBinary()
		if err != nil {
			return nil, err
		}
		enc.RawTransaction = rawTx
	} else {
		enc.RawTransaction = c.RawTransaction
	}
	enc.MaxBlockNumber = (*hexutil.Big)(c.MaxBlockNumber)
	enc.Preferences.Fast = c.Fast
	return json.Marshal(&enc)
}

// SendPrivateTransaction sends a private transaction to the Flashbots relay.
func SendPrivateTransaction(r *SendPrivateTransactionRequest) core.CallerFactory[common.Hash] {
	return &sendPrivateTransactionFactory{params: r}
}

type sendPrivateTransactionFactory struct {
	// args
	params *SendPrivateTransactionRequest

	// returns
	returns *common.Hash
}

func (f *sendPrivateTransactionFactory) Returns(txHash *common.Hash) core.Caller {
	f.returns = txHash
	return f
}

func (f *sendPrivateTransactionFactory) CreateRequest() (rpc.BatchElem, error) {
	return rpc.BatchElem{
		Method: "eth_sendPrivateTransaction",
		Args:   []any{f.params},
		Result: f.returns,
	}, nil
}

func (f *sendPrivateTransactionFactory) HandleResponse(elem rpc.BatchElem) error {
	if err := elem.Error; err != nil {
		return err
	}
	return nil
}
