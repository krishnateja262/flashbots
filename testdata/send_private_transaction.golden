> {"jsonrpc":"2.0","id":1,"method":"eth_sendPrivateTransaction","params":[{"tx":"0x00","maxBlockNumber":"0x98967f","preferences":{"fast":true}}]}
< {"jsonrpc":"2.0","id":1,"result":"0x45df1bc3de765927b053ec029fc9d15d6321945b23cac0614eb0b5e61f3a2f2a"}
